src/cts/src/HTreeBuilder.cpp

```diff
@@ -363,103 +346,463 @@ void adjustToplevelTopology(Point<double>& a,
+    double diff = abs(dist - targetDist);
+    // clang-format off
+    if (logger_->debugCheck(utl::CTS, "legalizer", 3)) {
+      logger_->report("      Loc {}: curr dist={:0.3f} target dist={:0.3f} sink"
+		      " dist={:0.3f}", loc, dist, targetDist,
+		      weightedDistance(loc, branchPoint, sinks));
+    }
+    // clang-format on
+    if (diff < minDiff) {
+      minDiff = diff;
+      best = loc;
     }
   }
 
+  return adjustBestLegalLocation(
+      targetDist, best, parentPoint, sinks, x1, y1, x2, y2, scalingFactor);
+}
+
+// Adjust buffer location in two steps:
+// step1: try moving buffer to existing blockage boundary (less expensive)
+// step2: try moving buffer beyond existing blockage boundary (more expensive)
+// In both steps, the first priority is to match target distance from current
+// point to parent point.  The second priority is to lower the weighted sink
+// distance
+Point<double> HTreeBuilder::adjustBestLegalLocation(
+    double targetDist,
+    Point<double> currLoc,
+    Point<double> parentPoint,
+    const std::vector<Point<double>>& sinks,
+    double x1,
+    double y1,
+    double x2,
+    double y2,
+    int scalingFactor)
+{
+  if (fuzzyEqual(targetDist, currLoc.computeDist(parentPoint))) {
+    return currLoc;
+  }
+
+  // try moving along blockage boundary
+  Point<double> bestLoc = currLoc;
+  if (adjustAlongBlockage(targetDist,
+                          currLoc,
+                          parentPoint,
+                          sinks,
+                          x1,
+                          y1,
+                          x2,
+                          y2,
+                          scalingFactor,
+                          bestLoc)) {
+    return bestLoc;
+  }
+
+  // try moving beyond blockage boundary
+  return adjustBeyondBlockage(
+      currLoc, parentPoint, targetDist, sinks, scalingFactor);
+}
+
+bool HTreeBuilder::adjustAlongBlockage(double targetDist,
+                                       Point<double> currLoc,
+                                       Point<double> parentPoint,
+                                       const std::vector<Point<double>>& sinks,
+                                       double x1,
+                                       double y1,
+                                       double x2,
+                                       double y2,
+                                       int scalingFactor,
+                                       Point<double>& bestLoc)
+{
+  Point<double> newLoc = currLoc;
+  // clang-format off
+  debugPrint(logger_, CTS, "legalizer", 3, "{} currDist={:0.3f} != "
+	     "targetDist={:0.3f}, adjustAlongBlockage...", currLoc,
+	     currLoc.computeDist(parentPoint), targetDist);
+  // clang-format on
+  double x = currLoc.getX();
+  double y = currLoc.getY();
+  double px = parentPoint.getX();
+  double py = parentPoint.getY();
+
+  Point<double> noBest(-1e6, -1e6);
+  bestLoc = noBest;
+  double bestSinkDist = 1e12;
```

Comment:
Perhaps using std::numeric_limits instead of magic numbers? I believe it will make the code more readable.
```
std::numeric_limits<double>::min()
std::numeric_limits<double>::max()
```

Reply:
Done.